Foreman facts uploading correctly merges facts

ares opened the pull request:
##### SUMMARY
Foreman facts callback does not work well when more facts are added from roles

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
foreman

##### ADDITIONAL INFORMATION
We upload facts for any "ok" event, given Foreman behavior, that removes all existing facts in Foreman of Ansible source. While setup phase upload all ansible_* modules, if next play introduces new fact, it deletes all facts from step one and adds just newly added facts. Also sending update for each play is not performance heavy, it's better to build the final set in memory and then send at once.

```paste below
there's no visible change in fact
```

ares committed:
Foreman facts uploading correctly merges facts

ares commented:
@ekohl mind to give it a quick read since you were debugging the issue with me? thanks

ekohl commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,34 @@ def _ssl_verify(self):
+            ansible_facts = {}
+            for _, result in results:
+                ansible_facts.update(result.get('ansible_facts', {}))
+            data = {}
```
Probably cleaner to write something like:
```python
facts = {
    "name": host,                     
    "facts": {            
        "ansible_facts": ansible_facts,                        
        "_type": "ansible",            
        "_timestamp": datetime.now().strftime(self.TIME_FORMAT),
    },                                                          
}
```

ekohl commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,34 @@ def _ssl_verify(self):
+        for host, results in self.items.items():
+            ansible_facts = {}
+            for _, result in results:
+                ansible_facts.update(result.get('ansible_facts', {}))
```
I know I wrote this, but it can be easier to read (and perhaps slightly more efficient):
```python
for _, result in results:
    if 'ansible_facts' in result:
        ansible_facts.update(result['ansible_facts'])
```

bcoca commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,34 @@ def _ssl_verify(self):
+                                  verify=self.ssl_verify)
+                r.raise_for_status()
+            except requests.exceptions.RequestException as err:
+                print(str(err))
```
no prints, no str , use display and to_text

bcoca reviewed (changes requested):
This can end up creating a large memory baloon as it will be an additional cache of all host facts for each host in play. Since this seems to try to compensate the fact that there is no API call to append facts to a host in foreman I would suggest trying to use a fact cache plugin to populate this instead.

mattclay commented:
The sanity test failures are valid. The integration test failures are unrelated and can be ignored.

ares commented:
Thank you very much both for looking. I updated what you suggested. Regarding performance concern, it's valid but this PR doesn't make the existing behavior much worse. This covers the case when some tasks register additional facts, usually that's lower amount as for setup in the beginning. Changing the way how Foreman parses facts would be nice, but a lot of things we rely on today would no longer apply. I don't think we can modify it anytime soon.

I'll also look at sanity test failures.

ares commented:
Would it be please possible to cherry-pick this to stable-2.7, it is quite important fix for use with Foreman
I think it's ready_for_review

ares commented:
Thanks Brian! @ewoud, mind to take a look? As per bot help, your `shipit` should help :-)

ekohl reviewed (approved):
:shipit:

ekohl commented:
shipit

ekohl reviewed (approved):
shipit

ares commented:
reopening because of stale_ci flag

iNecas reviewed (approved):
Would be great to get this in, as without this, the callback is not very useful once you start using things like `set_facts`

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,37 @@ def _ssl_verify(self):
+
+            try:
+                r = requests.post(url=self.FOREMAN_URL + '/api/v2/hosts/facts',
+                                  data=json.dumps(facts),
```
```suggestion
                                  json=facts,
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -196,26 +207,24 @@ def send_reports(self, stats):
             except requests.exceptions.RequestException as err:
-                print(to_text(err))
+                self._display.warning('Sending report to Foreman at {url} failed for {host}: {err}'.format(
+                    host=host, err=to_text(err), url=self.FOREMAN_URL))
```
We should probably use native strings here:
```suggestion
                    host=host, err=to_native(err), url=self.FOREMAN_URL))
```

cc @abadger

abadger replied:
No.  On the controller (callback plugins run on the controller), always use text strings.

But that does mean you should mark the format string with a u"".  There's some information about when that is important here: https://anonbadger.wordpress.com/2016/01/05/python2-string-format-and-unicode/

webknjaz replied:
@abadger I checked the source of that `warning()` function and it seems like it concatenates input with native strings which concerns me.

abadger replied:
yeah, please fix warning and deprecated.  Those need to use text strings when they concatenate.  (They really should use text strings everywhere and if I was fixing it, I would change all the literal strings into text strings but the concatenation is only thing I saw that causes an actual problem due to the various ways that Python2 and Python3 handle string formatting).

The way warning and deprecated are now, they could either traceback or cause tracebacks inside of display() when used with non-ascii strings of the wrong time.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,37 @@ def _ssl_verify(self):
-            print(to_text(err))
+        for host, results in self.items.items():
+            ansible_facts = {}
+            for dummy, result in results:
```
It might make sense to move this to `append_result`. Otherwise, abstractions are leaking here + it creates too much nesting which is bad for readability.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,37 @@ def _ssl_verify(self):
+                "facts": {
+                    "ansible_facts": ansible_facts,
+                    "_type": "ansible",
+                    "_timestamp": datetime.now().strftime(self.TIME_FORMAT),
```
Are you sure that you're using the right timezone?

webknjaz replied:
I see that it's copy-pasted so you could probably fix this separately from this PR.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -134,26 +134,37 @@ def _ssl_verify(self):
+                                  headers=self.FOREMAN_HEADERS,
+                                  cert=self.FOREMAN_SSL_CERT,
+                                  verify=self.ssl_verify)
+                r.raise_for_status()
```
This is probably the only line which should be wrapped with try/except.

ekohl replied:
`requests.post` can raise various exceptions, like ConnectionError. See https://github.com/kennethreitz/requests/blob/9742da7f91f595796a412a8150aacc00191be039/requests/adapters.py#L497-L531

ekohl committed:
Use requests' built in json encoding

This requires requests 2.14 (released in 2014). We can also drop the
Accept header since we don't use the actual response. The Foreman API
doesn't even care about the Accept header.

ekohl committed:
Use unicode warning strings

ekohl committed:
Avoid redefining a built in function

ekohl committed:
Match function signatures to the parent class

ekohl committed:
Move build_log to a separate function

ekohl committed:
Prefer monotonic time

ekohl committed:
Factor out formatting the current timestamp

ekohl committed:
Simplify the report construction

ekohl committed:
Use a requests session

This allows connection reuse and simplifies the code.

ekohl committed:
Follow variable naming convention

ekohl committed:
Send timestamps as ISO formatted UTC times

ekohl commented:
In https://github.com/ekohl/ansible/tree/fix/facts_uploading I've rebased the branch and included lots of other cleanups. I haven't tested it yet. I was looking into writing integration tests, but I don't see any examples for a callback.

Should I open a new PR for this?

ares commented:
@ekohl you now have permissions on my repo, please push the changes here directly, thanks!

ekohl commented:
Updated. It's a much larger refactor now and I'll need to do some actual testing. Ideally I'd write some integration tests, but I'd need some pointers how to write a skeleton so I can throw some plays at it.

webknjaz commented:
@ekohl look at https://github.com/ansible/ansible/tree/devel/test/integration/targets/inventory_foreman
There's `runme.sh` which is automatically picked up by our test runner. So you should basically invoke Ansible with this callback plugin and probably check output or smth.
Here's some of simulators we've got: https://github.com/ansible?utf8=%E2%9C%93&q=test-container&type=&language=

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,169 @@
+
+def get_time():
+    """
+    Return the time for measuring duration. Prefers monotonic time but falls back to the regular
```
The first line of docstring is a title and should be one sentence fitting 70 chars as per PEP 257

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,169 @@
 from ansible.plugins.callback import CallbackBase
 
 
+def build_log(data):
```
docstring plz

ekohl replied:
Updated

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,169 @@
+    """
+    Return the current timestamp as a string to be sent over the network.
+    """
+    return datetime.utcnow().isoformat()
```
Is it the same for Python 2 and Python 3?

ekohl replied:
Yes they are the same, but not actually what I intended. Looks like `utcnow()` returns a naive datetime object so `isoformat()` doesn't actually output a timezone. Python 3 does have a `datetime.timezone.utc` but I don't see an easy equivalent in Python 2 so it's probably easier to use a manual format which includes `Z` to indicate it's UTC.

Edit: looks like there's others doing the same: https://github.com/ansible/ansible/blob/10c5e26ce83f2fbddd67acac17aa534eaeb788dc/lib/ansible/modules/cloud/amazon/ec2.py#L1229-L1232

webknjaz replied:
smth like `datetime.utcnow().replace(tzinfo=tz).isoformat()` maybe

ekohl replied:
I considered that, but I don't think pytz is a dependency you can count on and in Python 2 you don't have `datetime.timezone.utc`. The following could work, but is rather verbose:
```python
try:
  from datetime import timezone
except ImportError:  # Python 2
  return datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%fZ')
else:
  return datetime.utcnow().replace(tzinfo=timezone.utc).isoformat()
```

mattclay replied:
You may also be interested in my comment about UTC timezones here: https://github.com/ansible/ansible/pull/54792#discussion_r273107955

ares commented:
Before this turns into huge refactoring, we need small fix in order to make this callback work again. I'll help with testing, but I hope bigger refactoring can be done in separate PRs too :-)

ekohl commented:
I toyed with writing integration tests. I pushed them, but they don't have any verification step. It'll need I https://github.com/ansible/foreman-test-container/pull/5.

During testing I couldn't use the foreman inventory script without caching. This is probably a regression that Ansible 2.8 introduced in the cache refactor.

mattclay commented:
@ekohl Could you open an issue with details on the problem you encountered with the inventory script without caching?

ekohl commented:
I've updated the PR and now has assertions in the tests. To verify it actually works I've ran the same tests against a current Foreman nightly installation. This PR will probably still fail until the container image PR is merged.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,169 @@
+
+def get_time():
+    """
+    Return the time for measuring duration. Prefers monotonic time but falls back to the regular
```
Still invalid.

ekohl replied:
Updated

ares commented:
@webknjaz any more comments? seems like this is complete now

webknjaz commented:
@ares could you please ping me after May 14? I'm at PyCon now.

ares commented:
absolutely, I just hope we make it to Ansible 2.8

ares commented:
@webknjaz it's May 14, could you please take a look? We need to get this into 2.8. Thank you!

mattclay commented:
@ares Ansible 2.8.0 is [scheduled](https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_8.html) to be released on May 16th, so the earliest this could be included would be in Ansible 2.8.1.

ares commented:
Thanks @mattclay, when is 2.8.1 expected? I was trying to find the schedule for it, but failed.

mattclay commented:
2.8.1 will most likely be 2 - 3 weeks after 2.8.0 is released.

ares commented:
Thanks, please let us know if there's anything on our side we can do to have this merged.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
-                self._disable_plugin('The `requests` python module is too old.')
+            requests_version = requests.__version__.split('.')
+            if int(requests_version[0]) < 2 or int(requests_version[1]) < 14:
+                self._disable_plugin(u'The `requests` python module is too old.')
```
It'd be nice to hint users about what version is not old.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
-            requests_major = int(requests.__version__.split('.')[0])
-            if requests_major < 2:
-                self._disable_plugin('The `requests` python module is too old.')
+            requests_version = requests.__version__.split('.')
```
You could convert things to int right here:
```suggestion
            requests_version = tuple(map(int, requests.__version__.split('.')))
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
-            if requests_major < 2:
-                self._disable_plugin('The `requests` python module is too old.')
+            requests_version = requests.__version__.split('.')
+            if int(requests_version[0]) < 2 or int(requests_version[1]) < 14:
```
And then this check would be way more readable:
```suggestion
            if requests_version < (2, 14):
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
         self.disabled = True
         if msg:
-            self._display.warning(msg + ' Disabling the Foreman callback plugin.')
+            self._display.warning(msg + u' Disabling the Foreman callback plugin.')
```
We should probably explicitly `to_text` it
```suggestion
            self._display.warning(to_text(msg) + u' Disabling the Foreman callback plugin.')
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
-    def _ssl_verify(self):
-        if self.FOREMAN_SSL_VERIFY.lower() in ["1", "true", "on"]:
+    def _ssl_verify(self, option):
+        if option.lower() in ["1", "true", "on"]:
```
How about:
```suggestion
        from ansible.module_utils.parsing.convert_bool import boolean as to_bool

        try:
            verify = to_bool(option)
        except TypeError:  # it wasn't a boolean value
            verify = option  # Set to a CA bundle:
        finally:
            if verify is False:  # is only set to bool if try block succeeds
                requests.packages.urllib3.disable_warnings()
                self._display.warning(
                    u"SSL verification of %s disabled" % self.foreman_url,
                )
            return verify
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
+                },
+            }
+
+            url = self.foreman_url + '/api/v2/hosts/facts'
```
You reassign this var on each loop iteration, please move it outside.
```suggestion
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
+                response.raise_for_status()
+            except requests.exceptions.RequestException as err:
+                self._display.warning(u'Sending facts to Foreman at {url} failed for {host}: {err}'.format(
+                    host=host, err=to_text(err), url=self.foreman_url))
```
Explicit is better than what's happening implicitly:
```suggestion
                    host=to_text(host), err=to_text(err), url=to_text(self.foreman_url)))
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
-            # could make a GET request to get the Foreman version & do
-            # this automatically.
+
+            url = self.foreman_url + '/api/v2/config_reports'
```
Same here: move url outside of the loop:
```suggestion
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
             except requests.exceptions.RequestException as err:
-                print(to_text(err))
+                self._display.warning(u'Sending report to Foreman at {url} failed for {host}: {err}'.format(
+                    host=host, err=to_text(err), url=self.foreman_url))
```
```suggestion
                    host=to_text(host), err=to_text(err), url=to_text(self.foreman_url)))
```

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
+        value = result._result
+        self.items[host].append((name, value))
+        if 'ansible_facts' in value:
+            self.facts[host].update(value['ansible_facts'])
```
Maybe?
```suggestion
        self.facts[host].update(value.get('ansible_facts', {}))
```

ekohl replied:
I don't like that in this case. It's a defaultdict which means that if no facts are collected in this run at all, the host will still end up in `self.facts`. The result is that it will `POST` an empty hash to Foreman which removes all facts for that host. This version avoids that by first checking it.

webknjaz commented:
@areian @ekohl I've added some comments needing to be addressed. Other than that — looks fine :)


Oh, and fix that sanity check linter offense.

webknjaz commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -80,154 +80,173 @@
+        return time.monotonic()
+    except AttributeError:
+        return time.time()
+
```
```console
04:24 ERROR: Found 1 pep8 issue(s) which need to be resolved:
04:24 ERROR: lib/ansible/plugins/callback/foreman.py:119:1: E302 expected 2 blank lines, found 1 (100%)
```
```suggestion


```

ekohl committed:
Apply suggestions from code review

Co-Authored-By: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

ekohl commented:
Updated with commits to address feedback. I think I addressed all comments but let me know if I missed anything.

ares commented:
@webknjaz could you please take one more look?

abadger commented on lib/ansible/plugins/callback/foreman.py:
```diff
@@ -77,157 +77,183 @@
         if msg:
-            self._display.warning(msg + ' Disabling the Foreman callback plugin.')
+            self._display.warning(to_text(msg) + u' Disabling the Foreman callback plugin.')
         else:
```
Note, the advice to use to_text() here was incorrect.  to_text should be used if the msg is not text where the msg enters the system as a whole.  It should not be used where msg is used (unless it is a non-string type until then.  But with a name like msg, it should be a string type when passed in here)

ekohl replied:
So it should be `self._display.warning(msg + u' ...')`?

abadger replied:
Yes.  And then you look at where _display_plugin is called to make sure that msg is text type (unicode on python2; str on python3).  If there are variables instead of literal strings at that call site, you go back further to see if those variables are text type.  (Note that once you trace the variable back to an Ansible API, it should be text type unless it is explicitly documented as returning byte strings.  If it is an Ansible API and it does not return text type, that's a bug in the Ansible API).

abadger replied:
So yeah, the callers that I see are fine.  You should be able to simply remove to_text() here.

abadger replied:
The part of my comment about "unless it is a non-string type until then" means that you would use to_text if you have something like an exception object and were trying to pass that to display.warning.  Something like this:
``` python
try:
    1/0
except ZeroDivisionError as e:
    display.warning(u'Problem in code.  Error: ' + to_text(e))
```

In that case you are transforming a ZeroDivisionError object into a string.  That's an approptriate place to use to_text().

ekohl commented:
@abadger I think I've removed the redundant `to_text`. Could you have another look?

ehelms commented:
shipit

ares commented:
shipit

ehelms commented:
@ares mind re-pushing to trigger tests? the failure seems false

ares commented:
it's actually @ekohl's PR now, but close/reopen could help

ares commented:
@abadger, could you please recheck or dismiss the old review?

ares commented:
I've rebased the branch on top of devel, could we please get this in? It's been a while and is important fix for facts uploading.

ares commented:
@AlanCoding sorry to ping you directly, is there something we can do to get this in? It had [shipit] for a while, it has one approval and it's opened for a very long time, yet it contains critical fix for facts importing.

AlanCoding commented:
the bot put back on the stale_ci label, so you might help to rebase again and squash commits. Also, if it's a critical fix that might be backported, you should look at adding a changelog fragment.

Those are only generic comments, I am not involved in merge decisions here. IRC and mailing list are valid places to ask. I will make a quick attempt to pass this along.

ares commented:
Thanks! rebasing right-away

danielmellado commented:
!waffling stale_ci
